sw/device/lib/testing/usb_testutils_streams.c

```diff
@@ -0,0 +1,488 @@
+static void rx_show(void *stream_v, dif_usbdev_rx_packet_info_t packet_info,
+                    dif_usbdev_buffer_t buf) {
+  usbdev_stream_t *s = (usbdev_stream_t *)stream_v;
+  usb_testutils_streams_ctx_t *ctx = s->ctx;
+  usb_testutils_ctx_t *usbdev = ctx->usbdev;
+  uint8_t data[0x100U];
+  size_t bytes_read;
+  CHECK_DIF_OK(dif_usbdev_buffer_read(usbdev->dev, usbdev->buffer_pool, &buf,
+                                      data, packet_info.length, &bytes_read));
+  LOG_INFO("rx_show packet of %u byte(s) - read %u", packet_info.length,
+           bytes_read);
+  buffer_dump(data, bytes_read);
+}
+
+// Returns an indication of whether a stream has completed its data transfer
+bool usb_testutils_stream_completed(const usb_testutils_streams_ctx_t *ctx,
+                                    uint8_t id) {
+  // Locate the stream context information
+  CHECK(id < USBUTILS_STREAMS_MAX);
+  const usbdev_stream_t *s = &ctx->streams[id];
+
+  return (s->tx_bytes >= s->transfer_bytes) &&
+         (s->rx_bytes >= s->transfer_bytes);
+}
+
+// Initialize a stream, preparing it for use
+void usb_testutils_stream_init(usb_testutils_streams_ctx_t *ctx, uint8_t id,
+                               uint8_t ep_in, uint8_t ep_out,
+                               uint32_t transfer_bytes,
+                               usbdev_stream_flags_t flags, bool verbose) {
+  // Locate the stream context information
+  CHECK(id < USBUTILS_STREAMS_MAX);
+  usbdev_stream_t *s = &ctx->streams[id];
+
+  // We need to be able to locate the test context given only the stream
+  // pointer within the strm_tx_done callback from usb_testutils
+  s->ctx = ctx;
+
+  // Remember the stream IDentifier and flags
+  s->id = id;
+  s->flags = flags;
+
+  // Remember whether verbose reporting is required
+  s->verbose = verbose;
+
+  // Extract a couple of convenient flags; from our perspective
+  s->sending = ((flags & kUsbdevStreamFlagRetrieve) != 0U);
+  s->generating = ((flags & kUsbdevStreamFlagCheck) != 0U);
+
+  // Not yet sent stream signature
+  s->sent_sig = false;
+
+  // Initialize the transfer state
+  s->tx_bytes = 0u;
+  s->rx_bytes = 0u;
+  s->transfer_bytes = transfer_bytes;
+
+  // Initialize the LFSR state for transmission and reception sides
+  // - we use a simple LFSR to generate a PRND stream to transmit to the USBPI
+  // - the USBDPI XORs the received data with another LFSR-generated stream of
+  //   its own, and transmits the result back to us
+  // - to check the returned data, our reception code mimics both LFSRs
+  s->tx_lfsr = USBTST_LFSR_SEED(id);
+  s->rxtx_lfsr = s->tx_lfsr;
+  s->rx_lfsr = USBDPI_LFSR_SEED(id);
+
+  // Packet size randomization
+  s->tx_buf_size = BUFSZ_LFSR_SEED(id);
+
+  // Set up the endpoint for IN transfers (TO host)
+  //
+  // Note: We install the rx_show handler to catch any misdirected data
+  // transfers
+  void (*rx)(void *, dif_usbdev_rx_packet_info_t, dif_usbdev_buffer_t) =
```

Comment:
Could you declare the handlers types in the header to improve readability?
i.e 
```c
typedef void (*usb_testutils_rx_handler)(void *, dif_usbdev_rx_packet_info_t, dif_usbdev_buffer_t);
```

Reply:
Thanks for the suggestion; I do intend to do that later, along with various other non-functional tidyups throughout usb_testutils.

I have made the status_t/TRY changes to the newly-added usb_testutils functions, but left the existing usb_testutils code unmodified by the same token; TBD in a subsequent PR